ops/storage.py

```diff
@@ -0,0 +1,263 @@
+
+    def save_notice(self, event_path, observer_path, method_name):
+        self._db.execute('INSERT INTO notice VALUES (NULL, ?, ?, ?)',
+                         (event_path, observer_path, method_name))
+
+    def drop_notice(self, event_path, observer_path, method_name):
+        self._db.execute('''
+            DELETE FROM notice
+             WHERE event_path=?
+               AND observer_path=?
+               AND method_name=?
+            ''', (event_path, observer_path, method_name))
+
+    def notices(self, event_path):
+        if event_path:
+            c = self._db.execute('''
+                SELECT event_path, observer_path, method_name
+                  FROM notice
+                 WHERE event_path=?
+                 ORDER BY sequence
+                ''', (event_path,))
+        else:
+            c = self._db.execute('''
+                SELECT event_path, observer_path, method_name
+                  FROM notice
+                 ORDER BY sequence
+                ''')
+        while True:
+            rows = c.fetchmany()
+            if not rows:
+                break
+            for row in rows:
+                yield tuple(row)
+
+
+class JujuStorage:
+    """"Storing the content tracked by the Framework in Juju.
+
+    This uses :class:`_JujuStorageBackend` to interact with state-get/state-set
+    as the way to store state for the framework and for components.
+    """
+
+    def __init__(self, backend: '_JujuStorageBackend'):
+        self._backend = backend
+
+    def close(self):
+        return
+
+    def commit(self):
+        return
+
+    def save_snapshot(self, handle_path: str, snapshot_data: typing.Any) -> None:
+        self._backend.set(handle_path, snapshot_data)
+
+    def load_snapshot(self, handle_path):
+        return self._backend.get(handle_path)
+
+    def drop_snapshot(self, handle_path):
+        self._backend.delete(handle_path)
+
+    def save_notice(self, event_path, observer_path, method_name):
+        notice_list = self.load_notice_list()
+        notice_list.append([event_path, observer_path, method_name])
+        self.store_notice_list(notice_list)
+
+    def drop_notice(self, event_path, observer_path, method_name):
+        notice_list = self.load_notice_list()
+        notice_list.remove([event_path, observer_path, method_name])
+        self.store_notice_list(notice_list)
+
+    def notices(self, event_path):
+        notice_list = self.load_notice_list()
+        if event_path:
+            notice_list = list(filter(lambda row: row[0] == event_path, notice_list))
+        for row in notice_list:
+            yield tuple(row)
+
+    def load_notice_list(self):
+        serialized_notices = self._backend.get("#notices#")
+        if serialized_notices is None:
+            return []
+        return serialized_notices
+
+    def store_notice_list(self, notice_list):
+        self.self._backend.set("#notices#", notice_list)
+
+
+class _SimpleLoader(yaml.SafeLoader):
```

Comment:
this is fine for a first iteration.

For a later iteration, maybe we can look at whether it's worth using `yaml.CSafeLoader` instead.